Applications/Spire/Include/Spire/OrderImbalanceIndicator/CachedOrderImbalanceIndicatorModel.hpp

```diff
@@ -0,0 +1,61 @@
+#ifndef SPIRE_CACHED_ORDER_IMBALANCE_INDICATOR_MODEL_HPP
+#define SPIRE_CACHED_ORDER_IMBALANCE_INDICATOR_MODEL_HPP
+#include <unordered_set>
+#include <boost/icl/continuous_interval.hpp>
+#include <boost/icl/interval_set.hpp>
+#include "Spire/OrderImbalanceIndicator/OrderImbalanceIndicatorModel.hpp"
+
+namespace Spire {
+
+  //! Represents an OrderImbalanceModel that caches loaded order imbalances
+  //! from a source model.
+  class CachedOrderImbalanceIndicatorModel :
+      public OrderImbalanceIndicatorModel {
+    public:
+
+      //! Constructs a CachedOrderImbalanceModel with a given source model.
+      /*
+        \param source_model The model supplying data to the cached model.
+      */
+      CachedOrderImbalanceIndicatorModel(
+        std::shared_ptr<OrderImbalanceIndicatorModel> source_model);
+
+      OrderImbalanceIndicatorModel::SubscriptionResult subscribe(
+        const boost::posix_time::ptime& start,
+        const boost::posix_time::ptime& end,
+        const OrderImbalanceSignal::slot_type& slot) override;
+
+    private:
+      struct OrderImbalanceHash {
+        std::size_t operator ()(const Nexus::OrderImbalance& imbalance) const;
+      };
+
+      struct Subscription {
+        OrderImbalanceSignal m_imbalance_signal;
+        boost::posix_time::ptime m_start_time;
+        boost::posix_time::ptime m_end_time;
+
+        Subscription(const boost::posix_time::ptime& start,
+          const boost::posix_time::ptime& end);
+      };
+
+      std::shared_ptr<OrderImbalanceIndicatorModel> m_source_model;
+      std::unordered_set<Nexus::OrderImbalance, OrderImbalanceHash>
```

Comment:
Can you use a ```vector``` sorted by timestamp instead of an ```unordered_set```? Looks like every subscription involves iterating over this entire set when it could be possible to calculate the index of the first imbalance, last imbalance, and return everything inbetween.

Reply:
I was trying to avoid having to write the code to merge the results because it involves finding the parts of the data that overlap and removing them.

Would a set work? I can't remember why I settled on an unordered_set, but if a set works it would provide the best combination of both; simpler merging and logarthmic searching.

Reply:
Yes you will need to remove duplicates. An ```std::set``` won't work unfortunately.

Reply:
What will be the secondary sort criteria if two imbalances share the same timestamps?

Reply:
You don't need a secondary sort criteria luckily.